did-sdk/src/ion/ion.ts

```diff
@@ -0,0 +1,340 @@
+        ? op
+        : last
+    }, this._ops[0])
+  }
+
+  getOperation(index: number): Op {
+    return this._ops[index]
+  }
+
+  getSuffix() {
+    return this.getURI('short').split(':').pop() || ''
+  }
+
+  assertActive() {
+    const lastOp = this.getLastOperation()
+    if (lastOp?.operation === 'deactivate') {
+      throw new Error('Cannot perform further operations on a deactivated DID')
+    }
+  }
+
+  private _createLongForm() {
+    // TODO
+    // this code is based on ion-tools, but is it correct?
+    // the longform DID is an encoding of the current state
+    // shouldn't that mean it needs to use the most recent op
+    // and not just the genesis?
+    // -prf
+    const create = this.getOperation(0) as OpCreate
+    if (!create) throw new Error('DID not created')
+    this._longForm =
+      this._longForm ||
+      IonSdk.IonDid.createLongFormDid({
+        recoveryKey: create.recovery.publicJwk,
+        updateKey: create.update.publicJwk,
+        document: create.content,
+      })
+  }
+
+  private async _resolveDidDoc() {
+    this._didDoc = (
+      await resolve(this.getURI(), this._ionResolveEndpoint)
+    ).didDoc
+  }
+
+  private async _submitAnchorRequest(body: any) {
+    // @ts-ignore the d.ts is wrong for the pow sdk -prf
+    return ProofOfWorkSDK.submitIonRequest(
+      this._ionChallengeEndpoint,
+      this._ionSolutionEndpoint,
+      JSON.stringify(body),
+    )
+  }
+
+  // read api
+  // =
+
+  getURI(form = 'long'): string {
+    this._createLongForm()
+    if (!this._longForm) throw new Error('DID not created')
+    return !form || form === 'long'
+      ? this._longForm
+      : this._longForm.split(':').slice(0, -1).join(':')
+  }
+
+  // writable api
+  // =
+
+  async create(doc: CreateParams, options: KeyParams): Promise<OpCreate> {
+    this.assertActive()
+    const op: OpCreate = {
+      operation: 'create',
+      content: doc,
+      recovery: await generateKeyPair(options.keyType, options),
+      update: await generateKeyPair(options.keyType, options),
+    }
+    const reqBody = IonSdk.IonRequest.createCreateRequest({
+      recoveryKey: op.recovery.publicJwk,
+      updateKey: op.update.publicJwk,
+      document: op.content,
+    })
+    await this._submitAnchorRequest(reqBody)
+    this._ops.push(op)
+    this._createLongForm()
+    await this._resolveDidDoc()
+    return op
+  }
+
+  async update(params: UpdateParams, options: KeyParams): Promise<OpUpdate> {
+    this.assertActive()
+    const op: OpUpdate = {
+      operation: 'update',
+      content: params,
+      previous: this.getPreviousOperation('update'),
+      update: await generateKeyPair(options.keyType, options),
+    }
+    if (!op.previous.update) {
+      throw new Error('Update key not found on previous ops')
+    }
+    const reqBody = await IonSdk.IonRequest.createUpdateRequest({
+      didSuffix: this.getSuffix(),
+      signer: IonSdk.LocalSigner.create(op.previous.update.privateJwk),
```

Comment:
where is this private key getting persisted to? It looks like it's just memory in the tests

Reply:
This is the `serialize()` API which did:key and did:ion provide. You're supposed to store the output of those methods and then you can load them with `inst()`.

For did:ion, it seems you want to store a local copy of the operations. It might be sufficient to only store the latest keypairs but I haven't explored that.